dask_expr/_merge.py

```diff
@@ -49,6 +49,10 @@ class Merge(Expr):
         "shuffle_backend": None,
     }
 
+    def __init__(self, *args, _precomputed_meta=None, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._precomputed_meta = _precomputed_meta
```

Comment:
I think I would strongly prefer if we use a global key-value cache in the same way we cache dataset info for parquet. In fact, we should probably formalize this caching approach to avoid repeating the same kind of logic in multiple places.

It seems like a unique meta depends on a token like...

```python
    @functools.cached_property
    def _meta_cache_token(self):
        return _tokenize_deterministic(
            self.left._meta,
            self.right._meta,
            self.how,
            self.left_on,
            self.right_on,
            self.left_index,
            self.right_index,
            self.suffixes,
            self.indicator,
        )
```

If `self.left._meta` or `self.right._meta` were to change (due to column projection), we would need to recalculate `meta` anyway. However, if the `Merge` object was responsible for pushing down the column projection, we could always update the cache within the simplify logic (since we would already know how the meta needs to change).

Reply:
Fwiw I am also not too big of a fan of relying on meta in hashes, there are too many things in pandas that might mutate this unexpectedly, which would break this

Reply:
To clarify, I don't really mind if we implement stand-alone caching logic in `_merge.py` for now. The thing I'm unsure about in this PR is that we are overriding `__init__` so that we can effectively cache `_meta` without adding `_meta` as an official operand.

It may be the case that this is exactly how we should be attacking this problem in `Merge` (and maybe everywhere). For example, maybe we will eventually have a special `known_meta=` kwarg in `Expr`, which **all** expression objects could leverage. However, since it is not a proper operand, this mechanism feels a bit confusing and fragile to me.

>The slowdown in merge goes back to the nonempty meta objects, not the actual computation on empty objects.

I don't think I understand your point here. Either way we are effectively caching the output of `_meta`, no?

>some of the operations in Lower have side effects, which makes adjusting the meta objects of left and right bothersome and complicated.

I don't see how this is any different for `_precomputed_meta`? In any case where you are confident defining `_precomputed_meta`, you could also just add the "lowered" object to the global cache before returning it.

Reply:
>Fwiw I am also not too big of a fan of relying on meta in hashes, there are too many things in pandas that might mutate this unexpectedly, which would break this

Interesting. I'd say that should be a fundamental concern for dask-expr in then. What would be then most reliable way to cache the schema?

Reply:
> I don't think I understand your point here. Either way we are effectively caching the output of _meta, no?

Yes but the nature of the slowdown makes me think that we need it only in merge and not in other places as of now. I am open to rewriting this here as well if this turns out differently.

> I don't see how this is any different for _precomputed_meta? In any case where you are confident defining _precomputed_meta, you could also just add the "lowered" object to the global cache before returning it.

pandas has some caveats that might change your dtype in meta but not on the actual df. Relying on the initial meta seems saver to me. But this might also be totally wrong.

> To clarify, I don't really mind if we implement stand-alone caching logic in _merge.py for now. The thing I'm unsure about in this PR is that we are overriding __init__ so that we can effectively cache _meta without adding _meta as an official operand.

@mrocklin and I chatted offline and landed on this solution. One motivating factor was the last part here: https://github.com/dask-contrib/dask-expr/pull/284#discussion_r1310193998

Reply:
> Interesting. I'd say that should be a fundamental concern for dask-expr in then. What would be then most reliable way to hash the schema?

I don't have a good answer for that. Meta is still the best bet, but it has some limitations. This will get more stable in the future since we are deprecating all of these caveats at the moment.

Reply:
>Yes but the nature of the slowdown makes me think that we need it only in merge and not in other places as of now. I am open to rewriting this here as well if this turns out differently.

Okay, I think you are are talking about the decision not to write general `Expr`-wide caching code here, which is all good with me. I was only thinking about the decision to use `_precomputed_meta` instead of a simple k/v cache.

Possible problems with the k/v cache approach:
- The meta-hashing issue you mentioned
- We would be keeping the cached meta in memory even after we need it (also a problem for parquet)

Possible problems with `_precomputed_meta`:
- I suppose we are breaking with convention a bit (seems okay)
- Any `substitute_parameters` call will drop the information, even if you aren't changing information that is relevant to meta

> One motivating factor was the last part here ...
Right, I agree that it would be a mistake to make `_precomputed_meta` a proper operand.

Reply:
We would also have to patch meta of left and right in the HashJoin layer because that adds a column

``substitute_parameters`` parameters is annoying, we could override but that's not great either.

Reply:
Sorry for adding all this discussion.  I do get that you are focusing on the `Merge`-specific `meta` issue at hand.

I'm just doing my best to keep the big picture in mind - It seems like we are going to keep running into cases where we would benefit from caching information outside the `Expr` object itself. Therefore, it would be nice if we could design a formal system where a collection of different caches can be managed in one place.

That said, I definitely don't think we need to do something like that right now.

>We would also have to patch meta of left and right in the HashJoin layer because that adds a column

Right, `HashJoinP2P._meta_cache_token` does need to drop the hash columns.